### ISSUE TITLE
Importing jobs non deterministic!

### ISSUE THREAD

**Issue by richiereynolds** [rundeck/rundeck#3336]
Issue type: Bug report/Enhancement Request

**My Rundeck detail**

* Rundeck version: 2.10.7-1
* install type: (rpm,deb,war,launcher?)rpm
* OS Name/version: linus RHEL 6.8
* DB Type/version: mysql? postgres? h2?.. mysql

**Expected Behavior**
Importing jobs from jar file or yaml should end up with exactly those jobs on the rundeck server.

**How to Reproduce**
We used to import all our jobs into new rundeck instances using a jar file however many times there would be issues with that e.g. some of the just imported jobs would not match what was in the jar and have errors in them that made them unable to run. On one server the import would work fine, on the next, it wouldn't.

So, we moved to importing from a directory of yaml files.
However after we loop through and import all the yaml files to a server successfully, we then export from that server and compare to the files and often find differences.
We then have to import the jobs that show differences (sometimes multiple times) until the exports match our files. The differences even change sometimes between these individual imports of the same file!
The differences will be for different jobs on different servers and even change between full import runs on the same server.
The differences can be job settings, options etc. sometimes to the extent that the job can't even be displayed properly on the rundeck GUI.

**Comment by richiereynolds** [comment]
We have about 200 very varied jobs and it's a completely random selection that fail to import correctly (although reporting success)  every time. Or sometimes they all import correctly. It'll be a different set of jobs running exactly the same thing twice. I've yet to find any rational of logical pattern to it. I'll send through some of the diffs we see after the import next time I do it.

**Comment by richiereynolds** [comment]
FYI here is how we're doing (and attributing success to) the imports:

`curl -kSsv --header "X-Rundeck-Auth-Token:$authToken" -F xmlBatch=@"$file" "https://$rdhost:4440/api/17/project/$project/jobs/import?format=$format&dupeOption=update&uuidOption=remove" | grep "<succeeded count='1'>"`

That look ok?

**Comment by richiereynolds** [comment]
As an aside, after we import a job, a period of time after that a new tag seems to be added to the job definition specifying the project name in the job ref steps e.g.
```
     - description: Checking site-specific files
       jobref:
         args: -Node ${node.name} -Label ${option.Label} -LabelToDiff ${option.LabelToDiff} -Rollback ${option.Rollback}
         group: INSTALL/CHECKS
         name: CheckSiteSpecificFiles
         nodeStep: 'true'
+        project: SIM
```
This doesn't stop anything working in this project but it makes the jobs non portable. If we export the set of jobs to be imported into a different server, with different project names, then it stops that working. Our current workaround is that we strip that line from the files after we export. Why does this line appear?

**Comment by richiereynolds** [comment]
Have some diffs for you now. We ran our script to loop around and import all jobs from files, then exported the jobs from that server and compared the output to the files. here are the diffs we saw:
```
diff -u -I '^\s*id:.*$' -I '^\s*uuid:.*$' -r /var/lib/rundeck/run/ash/jobs/INSTALL/AutoRollbackConfigurationOnNodes.yaml /tmp/ashJobs/INSTALL/AutoRollbackConfigurationOnNodes.yaml
--- /var/lib/rundeck/run/ash/jobs/INSTALL/AutoRollbackConfigurationOnNodes.yaml     2017-12-19 14:27:47.171099430 +0000
+++ /tmp/ashJobs/INSTALL/AutoRollbackConfigurationOnNodes.yaml    2018-05-03 11:36:22.177840243 +0100
@@ -51,9 +51,6 @@
     enforced: true
     name: PostDeployClose
     value: N
-    values:
-    - N
-    - Y
   - description: |-
       Automatically populated. The node filter that will be applied when running this job.<br>
       This can be copied into the filter on the <b>Nodes Tab</b> to check which nodes this job will run on.
diff -u -I '^\s*id:.*$' -I '^\s*uuid:.*$' -r /var/lib/rundeck/run/ash/jobs/INSTALL/DeployMaterials.yaml /tmp/ashJobs/INSTALL/DeployMaterials.yaml
--- /var/lib/rundeck/run/ash/jobs/INSTALL/DeployMaterials.yaml    2017-12-19 14:27:47.188099430 +0000
+++ /tmp/ashJobs/INSTALL/DeployMaterials.yaml   2018-05-03 11:36:27.909839828 +0100
@@ -67,8 +67,6 @@
     name: UserMaterialsTargetFile
     required: true
     value: data/Materials.csv
-    values:
-    - data/Materials.csv
   - description: TPUSER permissions source file in Perforce based from <code>P4GLOBALDIRSPATH</code> in the <code>P4GLOBALDIRS</code> depot.
     enforced: true
     name: UserPermissionsSourceFile
diff -u -I '^\s*id:.*$' -I '^\s*uuid:.*$' -r /var/lib/rundeck/run/ash/jobs/TASKMANAGER/ScheduleCronTimeRestoration.yaml /tmp/ashJobs/TASKMANAGER/ScheduleCronTimeRestoration.yaml
--- /var/lib/rundeck/run/ash/jobs/TASKMANAGER/ScheduleCronTimeRestoration.yaml      2018-05-01 11:25:28.287082976 +0100
+++ /tmp/ashJobs/TASKMANAGER/ScheduleCronTimeRestoration.yaml     2018-05-03 11:36:35.725839262 +0100
@@ -77,10 +77,9 @@
       If it's not provided, then a new FCRF will be created.
     name: FcrfId
   - description: |-
-      Reason for scheduling this task.<br/>
+      Optional reason for scheduling this task.<br/>
       The reason will appear on TPSCHEDULER FWI task audit page.
     name: Reason
-    required: true
   - description: Initial state that a task will appear in, on TPSCHEDULERFWI screen.
     enforced: true
     name: InitialState
diff -u -I '^\s*id:.*$' -I '^\s*uuid:.*$' -r /var/lib/rundeck/run/ash/jobs/TASKMANAGER/ScheduleDRComponentRollback.yaml /tmp/ashJobs/TASKMANAGER/ScheduleDRComponentRollback.yaml
--- /var/lib/rundeck/run/ash/jobs/TASKMANAGER/ScheduleDRComponentRollback.yaml      2018-05-01 11:25:28.287082976 +0100
+++ /tmp/ashJobs/TASKMANAGER/ScheduleDRComponentRollback.yaml     2018-05-03 11:36:35.920839248 +0100
@@ -65,25 +65,16 @@
     name: PreDeployClose
     required: true
     value: N
-    values:
-    - N
-    - Y
   - description: Should a server be started after deploying the new software?
     enforced: true
     name: PostDeployStart
     required: true
     value: Y
-    values:
-    - N
-    - Y
   - description: Should a server be closed again after starting the server?
     enforced: true
     name: PostDeployClose
     required: true
     value: N
-    values:
-    - N
-    - Y
   - description: |
       <table>
       <tr><th><b>Setting</b></th><th><b>Usage</b></th><th><b>Outcome</b></th>
@@ -92,11 +83,7 @@
     enforced: true
     name: SingleTask
     required: true
-    value: Choose
-    values:
-    - Choose
-    - N
-    - Y
+    value: N
   - description: |-
       Existing FCRF ID to be associated with this task.
       If it's not provided, then a new FCRF will be created.
diff -u -I '^\s*id:.*$' -I '^\s*uuid:.*$' -r /var/lib/rundeck/run/ash/jobs/TPMODEL/RemoveAttributes.yaml /tmp/ashJobs/TPMODEL/RemoveAttributes.yaml
--- /var/lib/rundeck/run/ash/jobs/TPMODEL/RemoveAttributes.yaml   2017-12-19 14:27:47.291099428 +0000
+++ /tmp/ashJobs/TPMODEL/RemoveAttributes.yaml  2018-05-03 11:36:33.092839452 +0100
@@ -19,9 +19,9 @@
   - description: Allow filtering of resources to a single component
     name: Component
     valuesUrl: http://localhost:1080/ash/cgi-bin/getComps.sh
-  - description: Allows filtering of resources to a single region.
-    name: Region
-    valuesUrl: http://localhost:1080/ash/getResourceNames.jsp?project=${job.project}&type=region
+  - description: Allow filtering of the resources to a single TP.
+    name: TP
+    valuesUrl: http://localhost:1080/ash/getResourceNames.jsp?project=${job.project}&type=TP&regionname=${option.Region.value}
   - description: Allow filtering of the resources to a single TP.
     name: TP
     valuesUrl: http://localhost:1080/ash/getResourceNames.jsp?project=${job.project}&type=TP&regionname=${option.Region.value}
```

We then imported each of those 5 jobs individually from the same files, exported again and ran the diff again and still one of the jobs had a difference:
```
diff -u -I '^\s*id:.*$' -I '^\s*uuid:.*$' -r /var/lib/rundeck/run/ash/jobs/TPMODEL/RemoveAttributes.yaml /tmp/ashJobs/TPMODEL/RemoveAttributes.yaml
--- /var/lib/rundeck/run/ash/jobs/TPMODEL/RemoveAttributes.yaml   2017-12-19 14:27:47.291099428 +0000
+++ /tmp/ashJobs/TPMODEL/RemoveAttributes.yaml  2018-05-03 11:39:29.048826715 +0100
@@ -22,9 +22,11 @@
   - description: Allows filtering of resources to a single region.
     name: Region
     valuesUrl: http://localhost:1080/ash/getResourceNames.jsp?project=${job.project}&type=region
-  - description: Allow filtering of the resources to a single TP.
-    name: TP
-    valuesUrl: http://localhost:1080/ash/getResourceNames.jsp?project=${job.project}&type=TP&regionname=${option.Region.value}
+  - description: The resource the attributes will be removed from.
+    enforced: true
+    name: Node
+    required: true
+    valuesUrl: http://localhost:1080/ash/getResourceNames.jsp?project=${job.project}&regionname=${option.Region.value}&tpname=${option.TP.value}&includeRegion=${option.Region.value}&includeTP=${option.TP.value}
   - description: The resource the attributes will be removed from.
     enforced: true
     name: Node
```
Then we ran the import again for that single jobs, exported again and finally saw no diffs.

**Comment by gschueler** [comment]
strange.. a few questions:

* are you using the Git plugin?
* do you have multiple rundeck servers using the same db (cluster mode)?
* do the servers have the same exact config when you export from one and import to another? (rundeck version, OS type/version, JDK vendor/version, DB type, etc.?)
* can you send me some of the jobs that seemed to have this problem (can email if you prefer)?

I'd also be interested in your script

**Comment by puremourning** [comment]
Richard, did you consider using the rd command line tool? We’ve not seen this issue on xit (the import hasn’t afaik ever done this sort of thing). What was the reason you used curl+api?

**Comment by richiereynolds** [comment]
@puremourning you'll be interested in this - we have moved to using rd jobs but have found several problems with that (apart from the obvious downside of having to be on the box the do the import):
1. It's incredibly slow, something like 20 times slower than using the API, our full suite of jobs now take half an hour rather than a couple of mins to import. Any idea why it's so slow in comparison?
2.We have to get the rd package installed on all servers, it doesn't comes with the rundeck rpm which isn't very handy.
3.We're still seeing the diffs issue! Will report more on this next week as we've a few more installs to do but we have seen it happen in dev.

Aside, I see that the unsolicited adding of "project" to jobRef steps which made jobs non portable has been  fixed, #2976  , which is great, that stops a lot of validation errors from our diff. However, we're too scared to upgrade rundeck to get it due to #2877  as restarting rundeck means we get job failures!
